File: matching.py
from typing import List, Dict, Any, Optional, Set

class UserCourse:
    def __init__(self, local_code: str, university_id: str):
        self.local_code = local_code
        self.university_id = university_id

class UserProfile:
    def __init__(
        self,
        gpa: float,
        completed_local_courses: List[UserCourse],
        english_scores: Optional[Dict[str, float]] = None,
        completed_waivers: Optional[List[str]] = None
    ):
        self.gpa = gpa
        self.completed_local_courses = completed_local_courses
        self.english_scores = english_scores or {}
        self.completed_waivers = completed_waivers or []

class DbCourseRequirement:
    def __init__(
        self,
        id: str,
        canonical_name: str,
        category: str,
        required_for_eligibility: bool,
        manual_review_required: bool,
        logic_group: Optional[str] = None
    ):
        self.id = id
        self.canonical_name = canonical_name
        self.category = category
        self.required_for_eligibility = required_for_eligibility
        self.manual_review_required = manual_review_required
        self.logic_group = logic_group

class DbEnglishRequirement:
    def __init__(
        self,
        id: str,
        waiver_type: str,
        minimum_grade: Optional[str] = None,
        verified: bool = False
    ):
        self.id = id
        self.waiver_type = waiver_type
        self.minimum_grade = minimum_grade
        self.verified = verified

class DbMajor:
    def __init__(
        self,
        id: str,
        name: str,
        gpa_requirement: Optional[float],
        course_requirements: List[DbCourseRequirement]
    ):
        self.id = id
        self.name = name
        self.gpa_requirement = gpa_requirement
        self.course_requirements = course_requirements

class DbUniversity:
    def __init__(
        self,
        id: str,
        name: str,
        majors: List[DbMajor],
        english_requirements: List[DbEnglishRequirement]
    ):
        self.id = id
        self.name = name
        self.majors = majors
        self.english_requirements = english_requirements

class DbRawCourseMapping:
    def __init__(self, university_id: str, local_course_code: str, canonical_name: str):
        self.university_id = university_id
        self.local_course_code = local_course_code
        self.canonical_name = canonical_name

def translate_local_courses(
    completed_local_courses: List[UserCourse],
    mappings: List[DbRawCourseMapping]
) -> Set[str]:
    completed_canonical = set()
    for course in completed_local_courses:
        for mapping in mappings:
            if (mapping.university_id == course.university_id and 
                mapping.local_course_code.strip().upper() == course.local_code.strip().upper()):
                completed_canonical.add(mapping.canonical_name)
                break
    return completed_canonical

def evaluate_eligibility(
    user_profile: UserProfile,
    targets: List[DbUniversity],
    mappings: List[DbRawCourseMapping]
) -> List[Dict[str, Any]]:
    completed_canonical = translate_local_courses(user_profile.completed_local_courses, mappings)
    results = []

    for uni in targets:
        for major in uni.majors:
            details = []
            satisfied_courses = []
            missing_courses = []
            manual_review_required = False

            # 1. GPA Check
            gpa_required = major.gpa_requirement
            gpa_satisfied = gpa_required is None or user_profile.gpa >= gpa_required

            # 2. English Requirements Check
            english_satisfied = True
            if len(uni.english_requirements) > 0:
                met_any = False
                for req in uni.english_requirements:
                    if req.waiver_type in user_profile.completed_waivers:
                        met_any = True
                        break
                    
                    if req.waiver_type == 'TOEFL' and 'toefl' in user_profile.english_scores and req.minimum_grade:
                        try:
                            min_score = float(req.minimum_grade)
                            if user_profile.english_scores['toefl'] >= min_score:
                                met_any = True
                                break
                        except ValueError:
                            pass
                    
                    if req.waiver_type == 'IELTS' and 'ielts' in user_profile.english_scores and req.minimum_grade:
                        try:
                            min_score = float(req.minimum_grade)
                            if user_profile.english_scores['ielts'] >= min_score:
                                met_any = True
                                break
                        except ValueError:
                            pass
                english_satisfied = met_any

            # 3. Course Requirements Check
            standalone_reqs = [r for r in major.course_requirements if not r.logic_group]
            grouped_reqs = {}
            for r in major.course_requirements:
                if r.logic_group:
                    grouped_reqs.setdefault(r.logic_group, []).append(r)

            # Process Standalone
            for req in standalone_reqs:
                is_satisfied = req.canonical_name in completed_canonical
                if req.manual_review_required:
                    manual_review_required = True

                if is_satisfied:
                    satisfied_courses.append(req.canonical_name)
                    details.append({
                        'canonicalName': req.canonical_name,
                        'category': req.category,
                        'status': 'SATISFIED',
                        'reason': f"이수한 교과목({req.canonical_name})이 매핑 조건에 부합합니다."
                    })
                else:
                    if req.required_for_eligibility:
                        missing_courses.append(req.canonical_name)
                        details.append({
                            'canonicalName': req.canonical_name,
                            'category': req.category,
                            'status': 'MANUAL_REVIEW' if req.manual_review_required else 'MISSING',
                            'reason': f"수동 확인 필요: 필수 이수 과목({req.canonical_name})의 이수 여부가 명확하지 않습니다." if req.manual_review_required else f"필수 과목 미이수: ({req.canonical_name}) 과목 이수가 필요합니다."
                        })
                    else:
                        details.append({
                            'canonicalName': req.canonical_name,
                            'category': req.category,
                            'status': 'MISSING',
                            'reason': f"권장 과목 미이수: ({req.canonical_name}) 과목은 필수 조건이 아닙니다."
                        })

            # Process Grouped (OR conditions)
            for group_name, reqs in grouped_reqs.items():
                satisfied_in_group = [req for req in reqs if req.canonical_name in completed_canonical]
                group_manual = any(req.manual_review_required for req in reqs)
                if group_manual:
                    manual_review_required = True
                
                is_group_required = any(req.required_for_eligibility for req in reqs)

                if len(satisfied_in_group) > 0:
                    for req in reqs:
                        is_completed = req.canonical_name in completed_canonical
                        if is_completed:
                            satisfied_courses.append(req.canonical_name)
                            details.append({
                                'canonicalName': req.canonical_name,
                                'category': req.category,
                                'status': 'SATISFIED',
                                'reason': f"선택 그룹({group_name}) 중 ({req.canonical_name}) 과목을 이수하여 조건이 충족되었습니다."
                            })
                        else:
                            details.append({
                                'canonicalName': req.canonical_name,
                                'category': req.category,
                                'status': 'SATISFIED',
                                'reason': f"선택 그룹({group_name})의 다른 과목이 이수되어 면제되었습니다."
                            })
                else:
                    if is_group_required:
                        option_names = " 또는 ".join(r.canonical_name for r in reqs)
                        for req in reqs:
                            missing_courses.append(req.canonical_name)
                            details.append({
                                'canonicalName': req.canonical_name,
                                'category': req.category,
                                'status': 'MANUAL_REVIEW' if req.manual_review_required else 'MISSING',
                                'reason': f"수동 확인 필요: 선택 그룹({group_name}) [{option_names}] 중 하나를 이수해야 합니다." if req.manual_review_required else f"필수 선택 누락: 선택 그룹({group_name}) [{option_names}] 중 최소 하나의 이수가 필요합니다."
                            })
                    else:
                        option_names = " 또는 ".join(r.canonical_name for r in reqs)
                        for req in reqs:
                            details.append({
                                'canonicalName': req.canonical_name,
                                'category': req.category,
                                'status': 'MISSING',
                                'reason': f"권장 선택 누락: 선택 그룹({group_name}) [{option_names}] 중 이수한 과목이 없습니다."
                            })

            # Check if any required course is missing
            required_missing = []
            for mc in missing_courses:
                for req in major.course_requirements:
                    if req.canonical_name == mc and req.required_for_eligibility:
                        required_missing.append(mc)
                        break

            is_eligible = gpa_satisfied and english_satisfied and (len(required_missing) == 0)

            results.append({
                'universityId': uni.id,
                'universityName': uni.name,
                'majorId': major.id,
                'majorName': major.name,
                'isEligible': is_eligible,
                'gpaSatisfied': gpa_satisfied,
                'englishSatisfied': english_satisfied,
                'gpaRequired': gpa_required,
                'gpaActual': user_profile.gpa,
                'satisfiedRequirementsCount': len(satisfied_courses),
                'totalRequirementsCount': len(major.course_requirements),
                'satisfiedCourses': satisfied_courses,
                'missingCourses': list(set(missing_courses)),
                'manualReviewRequired': manual_review_required,
                'details': details
            })

    return results

File: matching.ts
import { 
  UserProfile, 
  EligibilityResult, 
  EligibilityDetail, 
  UserCourse 
} from './types';

export interface DbCourseRequirement {
  id: string;
  canonicalName: string;
  category: string;
  requiredForEligibility: boolean;
  manualReviewRequired: boolean;
  logicGroup: string | null;
  officialSourceQuote?: string | null;
  sourceUrl?: string | null;
}

export interface DbEnglishRequirement {
  id: string;
  waiverType: string;
  minimumGrade?: string | null;
  rawQuote?: string | null;
  verified: boolean;
}

export interface DbMajor {
  id: string;
  name: string;
  gpaRequirement: number | null;
  courseRequirements: DbCourseRequirement[];
}

export interface DbUniversity {
  id: string;
  name: string;
  majors: DbMajor[];
  englishRequirements: DbEnglishRequirement[];
}

export interface DbRawCourseMapping {
  universityId: string;
  localCourseCode: string;
  canonicalName: string;
}

/**
 * Translates completed local courses to canonical course names based on mapping rules.
 */
export function translateLocalCourses(
  completedLocalCourses: UserCourse[],
  mappings: DbRawCourseMapping[]
): Set<string> {
  const completedCanonical = new Set<string>();
  
  for (const course of completedLocalCourses) {
    // Find a matching local mapping for the specific university
    const mapping = mappings.find(
      m => m.universityId === course.universityId && 
           m.localCourseCode.trim().toUpperCase() === course.localCode.trim().toUpperCase()
    );
    if (mapping) {
      completedCanonical.add(mapping.canonicalName);
    }
  }
  
  return completedCanonical;
}

/**
 * Checks transfer eligibility for a given user profile against target universities and majors.
 */
export function evaluateEligibility(
  userProfile: UserProfile,
  targets: DbUniversity[],
  mappings: DbRawCourseMapping[]
): EligibilityResult[] {
  const completedCanonical = translateLocalCourses(userProfile.completedLocalCourses, mappings);
  const results: EligibilityResult[] = [];

  for (const uni of targets) {
    for (const major of uni.majors) {
      const details: EligibilityDetail[] = [];
      const satisfiedCourses: string[] = [];
      const missingCourses: string[] = [];
      let manualReviewRequired = false;

      // 1. GPA Check
      const gpaRequired = major.gpaRequirement;
      const gpaSatisfied = gpaRequired === null || userProfile.gpa >= gpaRequired;

      // 2. English Requirements Check
      let englishSatisfied = true;
      if (uni.englishRequirements.length > 0) {
        // If there are English requirements, user must satisfy at least one waiver/grade requirement
        let metAnyRequirement = false;
        
        for (const req of uni.englishRequirements) {
          // Check waiver match
          if (userProfile.completedWaivers && userProfile.completedWaivers.includes(req.waiverType)) {
            metAnyRequirement = true;
            break;
          }
          
          // Check test scores based on waiver types (e.g., TOEFL, IELTS, SAT_ENGLISH)
          if (req.waiverType === 'TOEFL' && userProfile.englishScores?.toefl && req.minimumGrade) {
            const minScore = parseFloat(req.minimumGrade);
            if (!isNaN(minScore) && userProfile.englishScores.toefl >= minScore) {
              metAnyRequirement = true;
              break;
            }
          }
          if (req.waiverType === 'IELTS' && userProfile.englishScores?.ielts && req.minimumGrade) {
            const minScore = parseFloat(req.minimumGrade);
            if (!isNaN(minScore) && userProfile.englishScores.ielts >= minScore) {
              metAnyRequirement = true;
              break;
            }
          }
        }
        englishSatisfied = metAnyRequirement;
      }

      // 3. Course Requirements Check
      // Separate course requirements by logic groups (for OR conditions) and standalone
      const standaloneReqs = major.courseRequirements.filter(r => !r.logicGroup);
      const groupedReqs: Record<string, DbCourseRequirement[]> = {};
      
      for (const req of major.courseRequirements) {
        if (req.logicGroup) {
          if (!groupedReqs[req.logicGroup]) {
            groupedReqs[req.logicGroup] = [];
          }
          groupedReqs[req.logicGroup].push(req);
        }
      }

      // Process Standalone requirements
      for (const req of standaloneReqs) {
        const isSatisfied = completedCanonical.has(req.canonicalName);
        
        if (req.manualReviewRequired) {
          manualReviewRequired = true;
        }

        if (isSatisfied) {
          satisfiedCourses.push(req.canonicalName);
          details.push({
            canonicalName: req.canonicalName,
            category: req.category,
            status: 'SATISFIED',
            reason: `이수한 교과목(${req.canonicalName})이 매핑 조건에 부합합니다.`
          });
        } else {
          if (req.requiredForEligibility) {
            missingCourses.push(req.canonicalName);
            details.push({
              canonicalName: req.canonicalName,
              category: req.category,
              status: req.manualReviewRequired ? 'MANUAL_REVIEW' : 'MISSING',
              reason: req.manualReviewRequired 
                ? `수동 확인 필요: 필수 이수 과목(${req.canonicalName})의 이수 여부가 명확하지 않습니다.`
                : `필수 과목 미이수: (${req.canonicalName}) 과목 이수가 필요합니다.`
            });
          } else {
            details.push({
              canonicalName: req.canonicalName,
              category: req.category,
              status: 'MISSING',
              reason: `권장 과목 미이수: (${req.canonicalName}) 과목은 필수 조건이 아닙니다.`
            });
          }
        }
      }

      // Process Grouped requirements (OR conditions within the group)
      for (const [groupName, reqs] of Object.entries(groupedReqs)) {
        // Find which items in the group the user has satisfied
        const satisfiedInGroup = reqs.filter(req => completedCanonical.has(req.canonicalName));
        const groupManualReview = reqs.some(req => req.manualReviewRequired);
        
        if (groupManualReview) {
          manualReviewRequired = true;
        }

        const isGroupRequired = reqs.some(req => req.requiredForEligibility);

        if (satisfiedInGroup.length > 0) {
          // At least one course in the group is completed
          for (const req of reqs) {
            const isCompleted = completedCanonical.has(req.canonicalName);
            if (isCompleted) {
              satisfiedCourses.push(req.canonicalName);
              details.push({
                canonicalName: req.canonicalName,
                category: req.category,
                status: 'SATISFIED',
                reason: `선택 그룹(${groupName}) 중 (${req.canonicalName}) 과목을 이수하여 조건이 충족되었습니다.`
              });
            } else {
              details.push({
                canonicalName: req.canonicalName,
                category: req.category,
                status: 'SATISFIED', // Entire group is satisfied
                reason: `선택 그룹(${groupName})의 다른 과목이 이수되어 면제되었습니다.`
              });
            }
          }
        } else {
          // None of the courses in the group are completed
          if (isGroupRequired) {
            // Group options are added to missing courses
            const optionNames = reqs.map(r => r.canonicalName).join(' 또는 ');
            for (const req of reqs) {
              missingCourses.push(req.canonicalName);
              details.push({
                canonicalName: req.canonicalName,
                category: req.category,
                status: req.manualReviewRequired ? 'MANUAL_REVIEW' : 'MISSING',
                reason: req.manualReviewRequired
                  ? `수동 확인 필요: 선택 그룹(${groupName}) [${optionNames}] 중 하나를 이수해야 합니다.`
                  : `필수 선택 누락: 선택 그룹(${groupName}) [${optionNames}] 중 최소 하나의 이수가 필요합니다.`
              });
            }
          } else {
            // Optional group
            const optionNames = reqs.map(r => r.canonicalName).join(' 또는 ');
            for (const req of reqs) {
              details.push({
                canonicalName: req.canonicalName,
                category: req.category,
                status: 'MISSING',
                reason: `권장 선택 누락: 선택 그룹(${groupName}) [${optionNames}] 중 이수한 과목이 없습니다.`
              });
            }
          }
        }
      }

      // Final Eligibility Decision
      // Eligible if: GPA matches, English matches, and no *required* missing courses are present
      // Note: Grouped course options are added to missingCourses above. If any is missing,
      // the group is missing. But wait! For logic groups, we shouldn't fail if at least one was satisfied.
      // In the loop above, if satisfiedInGroup.length > 0, we did NOT add any of them to missingCourses.
      // If satisfiedInGroup.length === 0 and group is required, we added all option canonical names to missingCourses.
      // So checking `missingCourses.length === 0` correctly flags lack of completion.
      const coursesEligible = missingCourses.filter(mc => {
        // filter down to only courses that are strictly required for eligibility
        const req = major.courseRequirements.find(r => r.canonicalName === mc);
        return req ? req.requiredForEligibility : false;
      }).length === 0;

      const isEligible = gpaSatisfied && englishSatisfied && coursesEligible;

      results.push({
        universityId: uni.id,
        universityName: uni.name,
        majorId: major.id,
        majorName: major.name,
        isEligible,
        gpaSatisfied,
        englishSatisfied,
        gpaRequired,
        gpaActual: userProfile.gpa,
        satisfiedRequirementsCount: satisfiedCourses.length,
        totalRequirementsCount: major.courseRequirements.length,
        satisfiedCourses,
        missingCourses: Array.from(new Set(missingCourses)),
        manualReviewRequired,
        details
      });
    }
  }

  return results;
}
